packages/eslint-plugin/src/rules/thenable-in-promise-aggregators.ts

```diff
@@ -0,0 +1,176 @@
+      recommended: 'strict',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      inArray:
+        'Unexpected non-Thenable value in array passed to promise aggregator.',
+      arrayArg:
+        'Unexpected array of non-Thenable values passed to promise aggregator.',
+    },
+    schema: [],
+    type: 'problem',
+  },
+  defaultOptions: [],
+
+  create(context) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+
+    const aggregateFunctionNames = ['all', 'race', 'allSettled', 'any'];
+
+    function skipChainExpression<T extends TSESTree.Node>(
+      node: T,
+    ): T | TSESTree.ChainElement {
+      return node.type === AST_NODE_TYPES.ChainExpression
+        ? node.expression
+        : node;
+    }
+
+    function isPartiallyLikeType(
+      type: ts.Type,
+      predicate: (type: ts.Type) => boolean,
+    ): boolean {
+      if (isTypeAnyType(type) || isTypeUnknownType(type)) {
+        return true;
+      }
+      if (type.isIntersection() || type.isUnion()) {
+        return type.types.some(t => isPartiallyLikeType(t, predicate));
+      }
+      return predicate(type);
+    }
+
+    function isIndexableWithSomeElementsLike(
+      type: ts.Type,
+      predicate: (type: ts.Type) => boolean,
+    ): boolean {
+      if (isTypeAnyType(type) || isTypeUnknownType(type)) {
+        return true;
+      }
+
+      if (type.isIntersection() || type.isUnion()) {
+        return type.types.some(t =>
+          isIndexableWithSomeElementsLike(t, predicate),
+        );
+      }
+
+      if (!checker.isArrayType(type) && !checker.isTupleType(type)) {
+        const indexType = checker.getIndexTypeOfType(type, ts.IndexKind.Number);
+        if (indexType === undefined) {
+          return false;
+        }
+
+        return isPartiallyLikeType(indexType, predicate);
+      }
+
+      const typeArgs = type.typeArguments;
+      if (typeArgs === undefined) {
+        throw new Error(
+          'Expected to find type arguments for an array or tuple.',
+        );
+      }
+
+      return typeArgs.some(t => isPartiallyLikeType(t, predicate));
+    }
+
+    return {
+      CallExpression(node: TSESTree.CallExpression): void {
+        const callee = skipChainExpression(node.callee);
+        if (callee.type !== AST_NODE_TYPES.MemberExpression) {
+          return;
+        }
+
+        if (callee.computed) {
+          if (
+            callee.property.type !== AST_NODE_TYPES.Literal ||
+            typeof callee.property.value !== 'string' ||
+            !aggregateFunctionNames.includes(callee.property.value)
+          ) {
+            return;
+          }
+        } else if (!aggregateFunctionNames.includes(callee.property.name)) {
+          return;
+        }
+
+        const calleeType = services.getTypeAtLocation(callee.object);
+        if (
+          !isPromiseConstructorLike(services.program, calleeType) &&
+          !isPromiseLike(services.program, calleeType)
```

Comment:
[Performance] On the inside, both of these `is*Like` functions directly call to `isBuiltinSymbolLike`, which directly calls to `isBuiltinSymbolLikeRecurser`, which is a recursive function that roughly looks through all the node's types. So in the case of neither of these functions finding what they're looking for, they'll go through those steps twice.

This is also roughly the same code as `typeAtLocationIsLikePromise` from `prefer-promise-reject-errors`.

I think a good cleanup/followup/refactor would be to augment `isBuiltinSymbolLike` to take in an array or set of names that it checks for. That way we can do all that recursive stuff _once_ rather than _twice_ (one for each `symbolName`).

This is existing code (that I missed in a previous review!) so not a blocker at all IMO. Fine to file as an issue!

Reply:
Easy enough! I kept with the predicate pattern it already had going, which also has the benefit of giving a little more freedom to the caller without being too tedious.

I'll note that this was reaaallly confusing until I realized that I had to do some combination of `yarn` / `yarn build` for changes to actually be applied!

Reply:
Oh, I've been bit by that before. Do you think it'd be something useful to mention in docs? #8033 is an example of another recent augmentation in this line (not quite the same).

Reply:
When I was working on the type-utils package, I put a little stub to the `packages/type-utils/dist/index.js`:

```js
module.exports = require('jiti')(null, { interopDefault: true })('/path/to/typescript-eslint/packages/type-utils/src/index.ts')
```

This way, I don't have to rebuild the whole package every time I change something inside it. `jiti` just reads actual code from disk each time and transpiles in on the fly!

Perhaps we can add a script (or even use postinstall) that places stubs instead of package build artifacts when developing locally. I think it should improve DX

Ref: https://antfu.me/posts/publish-esm-and-cjs#stubbing

Reply:
I'd definitely vote for some kind of automatic local development flag, even if it's as simple as changing the launch configurations within vscode so that the jest rebuilds / double checks. I would have checked the docs next, but I had to first have the insight that it was actually not running the new code rather than me just doing something really silly in order to know to check the docs, at which point I was 99% of the way there as I then started trying yarn / yarn build commands.